feat: add optional stop before function to replayq:pop/2 options

kjellwinblad opened the pull request:
Two new optional options has been added to the replayq:pop/2 function:

* stop_before and,
* stop_before_input_accumulator.

The stop_before option is a function where the first argument is the current item, and the second argument is an accumulator (starts with the value of stop_before_input_accumulator if specified or #{} if stop_before_input_accumulator is not specified). The stop_before function should return true if the pop call should stop adding items before the current item. If the stop_before function does not return true, the returned value will be the accumulator argument to the next call of the stop_before function.

kjellwinblad committed:
feat: add optional stop before function to replayq:pop/2 options

Two new optional options has been added to the replayq:pop/2 function:
* stop_before and,
* stop_before_input_accumulator.

The stop_before option is a function where the first argument is the
current item, and the second argument is an accumulator (starts with the
value of stop_before_input_accumulator if specified or #{} if
stop_before_input_accumulator is not specified). The stop_before
function should return true if the pop call should stop adding items
before the current item. If the stop_before function does not return
true, the returned value will be the accumulator argument to the next
call of the stop_before function.

zmstone commented on src/replayq.erl:
```diff
@@ -226,13 +228,21 @@ append(#{config := #{seg_bytes := BytesLimit, dir := Dir} = Config,
+          #{
+            bytes_limit => bytes(),
+            count_limit => count(),
+            stop_before => stop_function(),
+            stop_before_input_accumulator => term()
```
```suggestion
            accumulator => {accumulator_func(), accumulator_initial_state()}.
```
default accumulator:
```
%% in this module
-export([default_accumulator/2]).
default_accumulator(Item, List) -> [Item | List].
```
batch stopper accumulator:

```
%% in the caller module
-export([collector/2]).
collector(Item, []) -> [Item];
collector(Item, [Last | _] = List) ->
  case can_be_batched(Item, Last) of
    true ->
      [Item | List];
    false ->
      false
end.
```

kjellwinblad replied:
Thanks for the suggestion. Why do you think the function should be called accumulator/accumulator_func? Its purpose is not to accumulate but to decide when to stop adding items.

kjellwinblad replied:
I think I understand what you mean now. You want this to be more general so that one can can accumulate items however one wants (for example, into something else than a list), right? I think this complicates things in a way that does not benefit our current use case and it will break the current interface since the list needs to be reversed in the end by the user if one does not add a special case. If we want to have a custom accumulator function we can add that as another option. However, I think making the stop_before and stop_before_input_accumulator options a single option makes sense so I will use this suggestion.

zmstone replied:
ok. can be a future enhancement as a `pop2` function.
this is beneficial when for example, we can encode a Kafka batch while popping, it gives more accuracy to the batch size.

currently the batch size is only an approximate estimation which can actually go over the limit due to the overhead bytes of each item, so we had to perform some oddly complicated compensation in estimation algorithms.

zmstone commented:
missing appup updates.

zmstone commented on src/replayq.erl:
```diff
@@ -307,59 +320,81 @@ transform(Id, [Item0 | Rest], Sizer, Count, Bytes, Acc) ->
-pop(Q, _Bytes, 0, AckRef, Acc) ->
+pop(Q, _Bytes, 0, AckRef, Acc, _StopFun, _StopFunAcc) ->
   Result = lists:reverse(Acc),
   ok = maybe_save_pending_acks(AckRef, Q, Result),
```
unrelated, but this should probably called after each pop call (not only when full-batch is collected).

kjellwinblad replied:
I think we can make another ticket/PR to fix this.

savonarola commented on src/replayq.erl:
```diff
@@ -24,6 +24,8 @@
 -type sizer() :: fun((item()) -> bytes()).
 -type marshaller() :: fun((item()) -> binary()).
+-type accumulator() :: term().
+-type stop_function() :: fun((item(), accumulator()) -> true | accumulator()).
```
Nit: seems that the only accumulating option that makes sense is cons (because we explicitly apply reverse to the acc), so why ask to implement it in the callback?
m.b. just 
```
-type need_accumulate() :: fun((item(), accumulator()) -> true | false).
```

kjellwinblad replied:
Not sure if I understand. The function is added because we want a way to get batches in emqx_resource_buffer_worker that only contain a certain kinds of requests. It would be very inefficient to scan all items pop'ed so far every time a new item is added to the list.

zmstone commented on src/replayq.erl:
```diff
@@ -226,13 +228,21 @@ append(#{config := #{seg_bytes := BytesLimit, dir := Dir} = Config,
 pop(Q, Opts) ->
   Bytes = maps:get(bytes_limit, Opts, ?DEFAULT_POP_BYTES_LIMIT),
   Count = maps:get(count_limit, Opts, ?DEFAULT_POP_COUNT_LIMIT),
+  StopFun = maps:get(stop_before, Opts, fun default_stop_function/2),
```
export this function and use `fun ?MODULE:default_stop_function/2`

kjellwinblad replied:
I guess this should be exported so that hot upgrade can work?

kjellwinblad replied:
Fixed

kjellwinblad committed:
fix: better structure for stop_before pop argument and readme text

Thanks @zmstone for suggesting combining the function and its initial
state into one argument.

kjellwinblad committed:
fix: update appup and export internal fun for hotupgradeability

kjellwinblad commented:
Appup file updated
